### DISCUSSION TITLE
Callbacks: Requirements and design

### DISCUSSION THREAD

**Discussion by nathanaelbosch** [probabilistic-numerics/probnum#438] [Ideas]
As we started discussing in #243, the currently implemented callbacks are as useful as we would like them to be (yet). For example, they can currently not be used to stochastically perturb the function estimates. So, the questions seem to be:
- What do we want to use the callbacks for? What capabilities do they need?
- How should they be implemented? What signature should they have? Or should values be stored in the solver itself, and then modified by accessing `self`?

**Comment by pnkraemer** [comment]
My main motivation for callbacks are being able to implement resets such as the ones in the Izhikevich model (https://www.izhikevich.org/publications/spikes.pdf) without rewriting `ODESolver.solve`. Another option could be perturbation schemes a la Conrad or Abdulle/Garegnagi, though @NinaEffenberger currently implements those by wrapping `step()` which I personally find cleaner. 

I have no strong opinion, especially on the latter point (the resets I am quite keen on via callbacks, unless you have a clearly better idea :) ).

**Comment by pnkraemer** [comment]
Maybe we can pick this up again here, and even broaden the scope of the discussion. Perhaps we can learn something from how @JonathanWenger uses callbacks in the linear algebra code, eventually leading to a probnum-wide convention? 

@JonathanWenger please feel no need to reply soon though. This is not urgent.

**Comment by JonathanWenger** [comment]
I ended up removing callbacks from the interface and adding a method to the class which implements the PN method and returns an iterator (in fact a generator). This can then be looped through and exposes all internals of the model. You can also use this to restart the solver from any state.

**Comment by pnkraemer** [comment]
Note for my future self: these guys (https://ode-solver.readthedocs.io/en/master/usage.html#classes-and-functions) implement ODE solvers with generators. In general, I find the idea very elegant, but I am not yet sure about speed.

**Comment by pnkraemer** [comment]
Hmmm, at least it does not seem slower... still not sure about advantages and disadvantages, but I kind of like the idea. 

![Screenshot from 2021-02-06 12-20-09](https://user-images.githubusercontent.com/45484438/107117016-93807b80-6877-11eb-96f5-5d660e46fd79.png)
